internal/gps/constraints.go

```diff
@@ -53,8 +53,13 @@ type Constraint interface {
 	// design goal of the system.
 	typedString() string
 
-	// equals returns true if the constraints are logically equivalent.
-	equals(Constraint) bool
+	// identical returns true if the constraints are identical.
+	//
+	// Identical Constraints behave identically for all methods defined by the
+	// interface. A Constraint is always identical to itself.
+	//
+	// Constraints serialized for caching are de-serialized into identical instances.
```

Comment:
It's more complicated than that. The encoding scheme is queued up for a future PR after this one (maybe the next, have to dig it out and rebase).